src/graph.h

```diff
@@ -13,26 +11,65 @@ namespace gs {
+  Graph(bool is_subgraph, torch::optional<torch::Tensor> col_ids,
+        torch::optional<torch::Tensor> row_ids, int64_t num_cols,
+        int64_t num_rows)
+      : is_subgraph_{is_subgraph},
+        num_cols_{num_cols},
+        num_rows_{num_rows},
+        col_ids_{col_ids},
+        row_ids_{row_ids} {}
   void LoadCSC(torch::Tensor indptr, torch::Tensor indices);
+  void LoadCSR(torch::Tensor indptr, torch::Tensor indices);
   void LoadCSCWithColIds(torch::Tensor column_ids, torch::Tensor indptr,
                          torch::Tensor indices);
   void SetCSC(std::shared_ptr<CSC> csc);
+  void SetCSR(std::shared_ptr<CSR> csr);
+  void SetCOO(std::shared_ptr<COO> coo);
+  void SetData(torch::Tensor data);
+  void CSC2CSR();
+  void CSR2CSC();
   std::shared_ptr<CSC> GetCSC();
-  c10::intrusive_ptr<Graph> ColumnwiseSlicing(torch::Tensor column_ids);
-  c10::intrusive_ptr<Graph> RowwiseSlicing(torch::Tensor row_ids);
+  std::shared_ptr<CSR> GetCSR();
+  std::shared_ptr<COO> GetCOO();
+  torch::optional<torch::Tensor> GetData();
+  int64_t GetNumRows();
+  int64_t GetNumCols();
+  c10::intrusive_ptr<Graph> ColumnwiseSlicing(torch::Tensor column_index);
+  c10::intrusive_ptr<Graph> RowwiseSlicing(torch::Tensor row_index);
   c10::intrusive_ptr<Graph> ColumnwiseSampling(int64_t fanout, bool replace);
   c10::intrusive_ptr<Graph> ColumnwiseFusedSlicingAndSampling(
-      torch::Tensor column_ids, int64_t fanout, bool replace);
-  torch::Tensor RowIndices(bool unique);
-  torch::Tensor AllIndices();
-  std::tuple<torch::Tensor, torch::Tensor, torch::Tensor> Relabel();
+      torch::Tensor column_index, int64_t fanout, bool replace);
+  torch::Tensor Sum(int64_t axis, int64_t powk);
+  torch::Tensor L2Norm(int64_t axis);
+  c10::intrusive_ptr<Graph> Divide(torch::Tensor divisor, int64_t axis);
+  c10::intrusive_ptr<Graph> Normalize(int64_t axis);
+  torch::Tensor AllValidNode();
+  torch::Tensor GetRows();
+  torch::Tensor GetCols();
+  torch::Tensor GetValidRows();
+  torch::Tensor GetValidCols();
+  torch::Tensor GetCOORows(bool is_original);
+  torch::Tensor GetCOOCols(bool is_original);
+  std::tuple<torch::Tensor, int64_t, int64_t, torch::Tensor, torch::Tensor,
+             torch::optional<torch::Tensor>, std::string>
+  Relabel();
   std::vector<torch::Tensor> MetaData();
 
-  void Print() const;
-
  private:
-  bool is_subgraph_;  // used for subgraph, stores node id in global graph
+  bool is_subgraph_;
+  int64_t num_cols_;  // total number of cols in a matrix
+  int64_t num_rows_;  // total number of rows in a matrix
   std::shared_ptr<CSC> csc_;
+  std::shared_ptr<CSR> csr_;
+  std::shared_ptr<COO> coo_;
+  torch::optional<torch::Tensor> data_;
+  torch::optional<torch::Tensor> col_ids_;  // column id in matrix
+  torch::optional<torch::Tensor> row_ids_;  // row id in matrix
+
+  std::tuple<torch::Tensor, torch::optional<torch::Tensor>,
+             torch::optional<torch::Tensor>>
+  PrepareDataForCompute(int64_t axis);
```

Comment:
Suggested changes:
- void CreateSparseFormat(int64_t axis) {}
- It does not return any thing. Operators should use csc_ or csr_ in the graph for computation. This design is flexible for an operator to use any data in the graph.

Reply:
Could you leave more details? I haven't got the point. If we do not return anything, what is this function for?

Reply:
For example, in `GraphNormalize(1)` (i.e., rowwise normalization), you call `CreateSparseFormat(1)` to create the CSR format of a graph. Then you can use `csr_->indptr`, `cse_->eids` and `data_` for computation.

Reply:
Oh I understand. While for this design, I have mainly two concerns.
1. Alongside this `CreateSparseFormat()` function, we still need a `PrepareDataForCompute()` operator to extract `indptr` and `e_ids` of either `CSC` or `CSR` to pass to `GraphSum/divide/normalize()` as parameters. Otherwise, we will need to write two identical `GraphSum()` with only different parameter `CSC` and `CSR` right?
2. Currently, only when dealing with graph computation, we decide whether to create `CSC`, `CSR` or `COO` by axis (0 or 1). For example, `relabel` on `CSR` format will invoke call to `CSR2COO` while its not dicided by axis.

Reply:
Add some changes on this.

Reply:
For 1, I'd suggest to have both GraphSum(axis=0/1) because we need to support them on both dimensions.
For 2, I cannot understand you point. From my understanding, `GraphRelabel` is an operator that returns some relabeled tensors.

Reply:
> For 1, I'd suggest to have both GraphSum(axis=0/1) because we need to support them on both dimensions. For 2, I cannot understand you point. From my understanding, `GraphRelabel` is an operator that returns some relabeled tensors.

GraphRelabel only support coo and csc format. It does not support csr format. Beacause in csr format, it forces srcnode(row_ids) to be relabeled beginning from 0 to xx. However, what we require is that  dstnode(col_ids) are relabeled beginning from 0. There are two ways to solve this problem. One is to generate a csr format, whose indptr contains srcnodes (row_ids) and dstnodes (col_ids) , but it costs a lot. Another way is doing csr2coo and then leverage coo to do relabel.  We choose the latter one.

Reply:
For 2, my point is that our provided `CreateSparseFormat(int64_t axis)` design is not that flexible, or as flexible as the previous design since for you saying that "This design is flexible for an operator to use any data in the graph.", I think only graph compute operators can use this `CreateSparseFormat(int64_t axis)` which is exactly the same as what `PrepareDataForCompute()` can do (in other operators like relabel, they don't have `axis` to pass to `CreateSparseFormat(int64_t axis)`).

BTW, it is a rather trivial point and I have already rewritten this. How do you think of the current version of this part?

Reply:
LGTM. `CreateSparseFormat` just make sure some sparse format exist when conducting some operators. It does not need to be used for all operators.

Go ahead and merge this PR when you feel it is ready.